src/libpriv/rpmostree-util.c

```diff
@@ -529,3 +529,74 @@ rpmostree_str_replace (const char  *buf,
 
   return g_regex_replace_literal (regex, buf, -1, 0, new, 0, error);
 }
+
+/* Like realpath(3), but ensures we stay within the given rootfs. We assume the
+ * rootfs is already made canonical. */
+
+char *
+rpmostree_realpath_within_rootfs (const char    *rootfs,
+                                  const char    *relpath,
+                                  GCancellable  *cancellable,
+                                  GError       **error)
+{
+  g_autofree char *path = NULL;
+  struct stat stbuf;
+
+  g_return_val_if_fail (rootfs != NULL, NULL);
+  g_return_val_if_fail (relpath != NULL, NULL);
+  g_return_val_if_fail (relpath[0] != '/', NULL);
+
+  path = g_build_filename (rootfs, relpath, NULL);
+
+  if (TEMP_FAILURE_RETRY (lstat (path, &stbuf)) != 0)
+    {
+      glnx_set_error_from_errno (error);
+      return NULL;
+    }
+
+  if (S_ISREG (stbuf.st_mode) || \
+      S_ISDIR (stbuf.st_mode))
+    {
+      g_autofree char *real = realpath (path, NULL);
```

Comment:
I believe it's still possible to walk outside of the root if one of the *intermediate* entries in the path is a symlink; e.g. `rootfs/usr/lib` -> `/usrlib` or something.

In the end, if we *have* to do this, the most obviously correct thing would be to run this code `chroot()ed` into the rootfs...except to do that as non-root which we eventually want would require using bwrap, which gets ugly.  But before that, to handle the Fedora situation I'd amend my suggestion to simply include `usr/lib/os.release.d/os-release-fedora` in the list.